### PR TITLE
 [WIP] Out-of-order: Extending Append method to include hints

### PR DESCRIPTION
**Pull request by Vanshikav123**
#11730 and #11765

### CODE REVIEW

**Review by bboreham — COMMENTED** [review]
Broadly solid PR, thanks!  Few points below; I think the scraping test needs work.

File: rules/manager_test.go [hunk 1]
```diff
@@ -1190,6 +1190,53 @@ func countStaleNaN(t *testing.T, st storage.Storage) int {
...
+	var stopped bool
+	ruleManager.start()
+	defer func() {
+		if !stopped {
+			ruleManager.Stop()
+		}
+	}()
+
+	file1 := "fixtures/rules2.yaml"
+	file2 := "fixtures/rules1.yaml"
```
[the remark this answers is not in the document]
Comment (Vanshikav123) [reply rules/manager_test.go:1223]: I will change these notations and will add comments

File: scrape/scrape_test.go [hunk 2]
```diff
@@ -81,6 +81,76 @@ func TestNewScrapePool(t *testing.T) {
...
+				metric: series.At().Labels(),
+				t:      t,
+				f:      v,
+			})
+		}
+		require.NoError(t, it.Err())
+	}
+	require.NoError(t, series.Err())
+
+	// Expected samples, only the in-order ones should be present.
```
[the remark this answers is not in the document]
Comment (Vanshikav123) [reply scrape/scrape_test.go:139]: we expect to discard out of order and only inorder is present....yaa name can be changed.. I can interchange checking it for Error and NoError..or something relevant configured / not-configured case.

File: util/teststorage/storage.go [hunk 3]
```diff
@@ -51,6 +51,8 @@ func NewWithError() (*TestStorage, error) {
 	opts.MaxBlockDuration = int64(24 * time.Hour / time.Millisecond)
 	opts.RetentionDuration = 0
 	opts.EnableNativeHistograms = true
+	//opts.OutOfOrderTimeWindow = 600000
```
[the remark this answers is not in the document]
Comment (Vanshikav123) [reply util/teststorage/storage.go:54]: Yaa a new function can be introduced ? or i can write in the test function itself?

**Pull request closed without merging**